src/chainstate/stacks/db/blocks.rs

```diff
@@ -2702,6 +2747,103 @@ impl StacksChainState {
 
         Ok(ret)
     }
+
+    pub fn will_admit_mempool_tx<R: Read>(&mut self, current_burn: &BurnchainHeaderHash, current_block: &BlockHeaderHash, fd: &mut R) -> Result<StacksTransaction, MemPoolRejection> {
+        // 1: it should parse as a tx.
+        let mut fd_counter = ReadCounter::from_reader(fd);
+        let tx = StacksTransaction::consensus_deserialize(&mut fd_counter)
+            .map_err(|e| MemPoolRejection::DeserializationFailure(e))?;
+
+        // 2: it must be validly signed.
+        StacksChainState::process_transaction_precheck(&self.config(), &tx)
+            .map_err(|e| MemPoolRejection::FailedToValidate(e))?;
+
+        // 3: it must pay a tx fee
+        let fee = tx.get_fee_rate();
+        let tx_size = fd_counter.read_count();
+
+        if fee < MINIMUM_TX_FEE || 
+           fee / tx_size < MINIMUM_TX_FEE_RATE_PER_BYTE {
+            return Err(MemPoolRejection::FeeTooLow(fee, cmp::max(MINIMUM_TX_FEE, tx_size * MINIMUM_TX_FEE_RATE_PER_BYTE)))
+        }
+
+        // 4: the account nonces must be correct
+        let (origin, payer) = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+            StacksChainState::check_transaction_nonces(conn, &tx)
+                .map_err(|e| MemPoolRejection::BadNonces(e))
+        })?;
+
+        // 5: the paying account must have enough funds
+        if fee as u128 > payer.stx_balance {
+            return Err(MemPoolRejection::NotEnoughFunds(fee as u128, payer.stx_balance))
+        }
+
+        // 6: payload-specific checks
+        match &tx.payload {
+            TransactionPayload::TokenTransfer(_addr, amount, _memo) => {
```

Comment:
In the case of a transfer, we could check that recipient and sender are both mainnet or testnet addresses?
I'm not sure.

Reply:
Yes, we should be doing this.  The address version byte must be a mainnet or testnet byte, depending on whether or not the chainstate is for mainnet or testnet.

Reply:
Moreover, if the `_addr` is a contract address, the contract must exist in order for this transaction to be accepted (it doesn't make sense to have a transaction send STX to a non-existent contract address IMHO).  Please insert a `TODO` comment for this.

Reply:
> Moreover, if the _addr is a contract address, the contract must exist in order for this transaction to be accepted (it doesn't make sense to have a transaction send STX to a non-existent contract address IMHO). Please insert a TODO comment for this.

I don't think this is necessary. While I don't know why someone would want to send STX to a contract that doesn't exist, I don't see any reason to prevent them from doing so. It seems analogous to sending STX to an normal address that hasn't been used yet -- i.e., we don't know if it actually corresponds to a private key, but we won't prevent someone from sending the transaction.

Reply:
How do you materialize the STX sent to a contract address that doesn't exist?  Do we materialize the contract with an empty code body?  Does the contract now exist?  Can I use this technique to send 1 microSTX in order to create a bunch of cheap contract names (and in doing so, get access to cheap data storage by encoding data in the contract name field)?

Reply:
> How do you materialize the STX sent to a contract address that doesn't exist? Do we materialize the contract with an empty code body? Does the contract now exist?

No -- this just inserts an account into the the account table with the contract's principal data.

> Can I use this technique to send 1 microSTX in order to create a bunch of cheap contract names (and in doing so, get access to cheap data storage by encoding data in the contract name field)?

Yes, though you are paying a transaction fee per byte for the token transfer. Note that this is also true of the memo field and standard addresses -- though you get fewer bytes in standard addresses.

Reply:
> No -- this just inserts an account into the the account table with the contract's principal data.

I'm not sure what this means.  Does this mean that I can send STX to a non-existent contract address, and then instantiate a contract at that address, and have that contract end up owning the STX (which it could transfer later)?

Reply:
> The address version byte must be a mainnet or testnet byte, depending on whether or not the chainstate is for mainnet or testnet.

Addressed in 105d66f3a

Reply:
Thanks!